### PR TITLE
Take 1 on large batch simulation with fewer GPUs

### PR DESCRIPTION
**Pull request by avital**

### CODE REVIEW

File: pytorch/main.py [hunk 1]
```diff
@@ -212,79 +212,104 @@ def train(train_loader, model, ema_model, optimizer, epoch, log):
...
+                assert len(model_out) == 2
+                assert len(ema_model_out) == 2
+                logit1, logit2 = model_out
+                ema_logit, _ = ema_model_out
+
+            ema_logit = Variable(ema_logit.detach().data, requires_grad=False)
+
+            if args.logit_distance_cost >= 0:
+                class_logit, cons_logit = logit1, logit2
+                res_loss = args.logit_distance_cost * residual_logit_criterion(class_logit, cons_logit) / minibatch_size / 2
```
Comment (lucasb-eyer) [inline pytorch/main.py:272]: You don't need to divide by two manually here, as `minibatch_size` is already half of what it originally is, since it's computed from `len(target_var)`, which is a chunk.

File: pytorch/main.py [hunk 2]
```diff
@@ -212,79 +212,104 @@ def train(train_loader, model, ema_model, optimizer, epoch, log):
...
+                meters.update('res_loss', res_loss.data[0])
+            else:
+                class_logit, cons_logit = logit1, logit1
+                res_loss = 0
+
+            class_loss = class_criterion(class_logit, target_var) / minibatch_size / 2
```
Comment (lucasb-eyer) [inline pytorch/main.py:278]: Same.

File: pytorch/main.py [hunk 3]
```diff
@@ -212,79 +212,104 @@ def train(train_loader, model, ema_model, optimizer, epoch, log):
...
+            meters.update('class_loss', class_loss.data[0])
+
+            ema_class_loss = class_criterion(ema_logit, target_var) / minibatch_size / 2
```
Comment (lucasb-eyer) [inline pytorch/main.py:281]: Same.

File: pytorch/main.py [hunk 4]
```diff
@@ -212,79 +212,104 @@ def train(train_loader, model, ema_model, optimizer, epoch, log):
...
+            meters.update('ema_class_loss', ema_class_loss.data[0])
+
+            if args.consistency:
+                consistency_weight = get_current_consistency_weight(epoch)
+                meters.update('cons_weight', consistency_weight)
+                consistency_loss = consistency_weight * consistency_criterion(cons_logit, ema_logit) / minibatch_size / 2
```
Comment (lucasb-eyer) [inline pytorch/main.py:287]: And same again.

File: pytorch/main.py (hunk 1 above)
Comment (lucasb-eyer) [reply pytorch/main.py:272]: Also, there is no `from __future__ ...` here, so make sure to run it with py3.

File: pytorch/main.py [hunk 5]
```diff
@@ -333,7 +358,7 @@ def validate(eval_loader, model, log, global_step, epoch):
         # compute output
         output1, output2 = model(input_var)
         softmax1, softmax2 = F.softmax(output1, dim=1), F.softmax(output2, dim=1)
-        class_loss = class_criterion(output1, target_var) / minibatch_size
+        class_loss = class_criterion(output1, target_var) / minibatch_size / 2
```
Comment (lucasb-eyer) [inline pytorch/main.py:361]: Again, here you don't need to divide by 2.

**Review by lucasb-eyer — COMMENTED** [review]
Only mistake I found was the division by two.

Actually, I don't understand why the code does the division by batch-size manually, instead of just setting `size_average=True` (the default) on the loss constructor? It doesn't make sense to me.

File: pytorch/main.py (hunk 1 above)
Comment (avital) [reply pytorch/main.py:272]: It's the other way around -- I am dividing by 2 to make it like it would be with the full minibatch size. Otherwise the effective learning rate would be wrong.

Comment (lucasb-eyer) [reply pytorch/main.py:272]: You are right, I was mistaken. Just for reference, here's the pseudo-math how I verified it.

```
grad = 0
original code: grad += sum_100 loss_i / 100  => 1 if all loss_i = 1
two-step: grad += sum_50 loss_i / 50 / 2  ; grad += sum_50 loss_i / 50 / 2  => 0.5 + 0.5 if all loss_i = 1
```
